### ISSUE TITLE
"apt changelog" broken for some Raspberry Pi-provided packages

### ISSUE THREAD

**Issue by andrum993** [raspberrypi/bookworm-feedback#6]
The `apt changelog` command fails to display the package changelog for some Raspberry Pi-provided packages in Bookworm.

Example:

`andrew@raspberrypi:~ $ apt changelog raspi-config
E: Failed to fetch changelog:/raspi-config.changelog  Changelog unavailable for raspi-config=20230821
`

The problem affects at least the following packages:

- raspi-config
- vlc
- raspberrypi-ui-mods

I would expect to see the same issue with other Raspberry Pi-provided packages, but not all of them. The Raspberry Pi-provided `kodi` package does not suffer from this problem, and neither does `wf-panel-pi`.

I did think that the problem was caused by apt ignoring the changelog if the file was named `changelog.gz`, and only displaying it if it was named `changelog.Debian.gz`. It seems that this is not the case, however, since for the `wf-panel-pi` package the changelog file is named `changelog.gz`, and it displays correctly. So something else is going on.

Linked forum thread: https://forums.raspberrypi.com/viewtopic.php?t=355062.

**Comment by XECDesign** [comment]
I don't think this is a Raspberry Pi OS issue. We haven't modified apt in any way and our packaging is relatively standard as far as changelogs go.

I was under the impression that these sorts of tools download the changelog from one of pre-defined locations. We don't store our changelogs anywhere outside of the package, so I would not have expected this to work in the first place.

If it has some fallback where it tried to extract the changelog from the .deb and fails, then it's a Debian issue.

**Comment by andrum993** [comment]
I don't know where the problem lies - I was just reporting the issue.

As to how `apt` works, I have read some of the relevant source code (at https://salsa.debian.org/apt-team/apt/-/blob/main/apt-pkg/acquire-item.cc), and it seems to have a fallback to allow it to read the changelog file from `/usr/share/doc`, which is where I was manually getting the changelogs from when `apt changelog` failed. I am assuming that all installed packages have their changelog deposited there when the package is installed.

**Comment by cillian64** [comment]
Had a little dig in to this, I don't think the problem is the changelog.gz files: On my system `apt changelog rpi-firefox-mods` works but `apt changelog rpi-chromium-mods` doesn't, and both of those have a /usr/share/doc/<PKGNAME>/changelog.gz file in the correct format.

Spent a while debugging apt and it seems that in the bad case it's deciding it can't find the specified package.  But from strace it is opening `/usr/share/doc/rpi-chromium-mods/changelog.gz`, but doesn't seem to be doing anything with it.

I did find that if you have a version of the package installed which is no longer available in the apt repo (eg because you haven't upgraded and we've pruned an older version of the package) then `apt changelog` seems to fail, but I also have a bad example where that isn't the case.

**Comment by cillian64** [comment]
Ok, the problem is essentially the same as https://unix.stackexchange.com/questions/672586/is-it-possible-to-supply-aptitude-with-changelogs-of-non-debian-packages

The packages for which `apt changelog` doesn't work are those which are trimmed  - only a partial changelog is installed to the system with the idea that you can fetch the full changelog from a metadata server.  You can see this if you run `zcat /usr/share/doc/rpi-chromium-mods/changelog.gz | tail`  But... as far as I know we don't run a metadata server and that's why apt is getting unhappy and giving a really unhelpful error message.

So the solution here is for us to either run a package metadata server, or to remove the "trimmed" markers from all our packages.  As a workaround you can just zcat the changelog file instead of using `apt changelog`.

**Comment by andrum993** [comment]
Just to add, since I hadn't actually mentioned it in my original message: this is a change from Bullseye.

On Raspberry Pi OS Bullseye `apt changelog` works as intended for all Raspberry Pi-provided packages that I have tested. On Raspberry Pi OS Bookworm, some don't.

It sounds like this is a result of an upstream change, i.e. something Debian have done. It's a minor issue, but one of those things that gets annoying after a while. For anyone else who finds this irritating, the following bash script can be used in place of a call to `apt changelog`:

```bash
#!/usr/bin/bash
DOCDIR=/usr/share/doc
PAGER=/usr/bin/more
LOGS=("changelog.Debian.gz" "changelog.gz")
if [ "$#" -ne 1 ] | [ "$#" -eq 0 ]; then
    echo "cl.sh: show changelog for an installed apt package"
    echo "usage: cl.sh packagename"
    echo "packagename = name of apt package to show changelog of"
    exit 0
elif [ "$#" -eq 1 ]; then
    for log in ${LOGS[@]}; do
        if [ -f $DOCDIR/$1/$log ]; then
            zcat $DOCDIR/$1/$log | $PAGER
            exit 0
        fi
    done
fi
```

`changelog.Debian.gz` is preferred over `changelog.gz`, since for one of the packages I tried, `pcmanfm`, the former had more recent changes done by Raspberry Pi, whereas the latter was from upstream and contained less recent changes.

HTH
